checkbox-support/checkbox_support/tests/test_timeout.py

```diff
@@ -137,38 +139,84 @@ def test_unpicklable_return_raises(self):
         def k():
             return lambda x: ...
 
-        with self.assertRaises(AttributeError):
+        with self.assertRaises(SystemExit):
             k()
 
-    @patch("checkbox_support.helpers.timeout.Process")
-    @patch("os.setsid")
-    def test_unpicklable_raise_raises(self, os_setid, process_type_mock):
+    def test_unpicklable_raise_raises(self):
         """
         The reason why this raises is that the timeout decorator pushes the
         function to another process. Trying to raise an un-picklable object
         will raise a pickle error.
         """
 
-        # this mocks process because else the coverage doesn't get the
-        # coverage
-        def init(*args, **kwargs):
-            process_type_mock.target = kwargs["target"]
-            process_type_mock.args = kwargs["args"]
-            process_type_mock.kwargs = kwargs["kwargs"]
-            return process_type_mock
-
-        def start():
-            return process_type_mock.target(
-                *process_type_mock.args, **process_type_mock.kwargs
-            )
-
-        process_type_mock.side_effect = init
-        process_type_mock.start = start
-        process_type_mock.is_alive.return_value = False
-
         @timeout(1)
         def k():
             raise ValueError(lambda x: ...)
 
         with self.assertRaises(SystemExit):
             k()
+
+    def test_timeout_kills_subprocess_tree_on_timeout(self):
+        """
+        This tests that the timeout decorator not only kills the direct child
+        (function under test) but also any sub-process it has spawned. This
+        is done because one could `subprocess.run` a long-lived process from
+        the function and it could cause mayhem (and block the test session as
+        Checkbox waits for all childs to be done)
+        """
+
+        def inner(pid_pipe):
+            pid_pipe.send(os.getpid())
+            pid_pipe.close()
+            time.sleep(1e4)
+
+        def outer(pid_pipe):
+            inner_p = multiprocessing.Process(target=inner, args=(pid_pipe,))
+            inner_p.start()
+            inner_p.join()
+
+        @timeout(0.1)
+        def f(pid_pipe):
+            outer_p = multiprocessing.Process(target=outer, args=(pid_pipe,))
+            outer_p.start()
+            outer_p.join()
+
+        read, write = multiprocessing.Pipe()
+        with self.assertRaises(TimeoutError):
+            f(write)
+        with self.assertRaises(OSError):
+            pid = read.recv()
+            # give the process a few ms to wind down
+            time.sleep(0.01)
```

Comment:
Is it because of the note in [the multiprocessing doc](https://docs.python.org/3/library/multiprocessing.html#pipes-and-queues):

> After putting an object on an empty queue there may be an infinitesimal delay before the queue’s [empty()](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.Queue.empty) method returns [False](https://docs.python.org/3/library/constants.html#False) and [get_nowait()](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.Queue.get_nowait) can return without raising [queue.Empty](https://docs.python.org/3/library/queue.html#queue.Empty).

?

Reply:
No, this is a pipe, not a queue! This sleep is because we sent a signal to the subprocess, it could take a few ms for it to be fully killed and removed from the valid pid list. Namely, 

we kill the tree -> tree starts winding down -> we recv the inner pid 

If we check right away, the tree may still be winding down, so we interrupt for an instant. Then we try to see if the inner process still exists.